pkg/operator/k8sutil/events.go

```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2021 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package k8sutil for Kubernetes helpers.
+package k8sutil
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+)
+
+type eventObject struct {
+	eventsReportedAt time.Time
+	eventCount       int
+}
+
+// EventReporter is custom events reporter type which allows user to limit the events
+type EventReporter struct {
+	recorder       record.EventRecorder
+	reportedEvents map[string]*eventObject
+
+	// report events x times where x is count
+	count int
+
+	// report events after x minutes
+	eventReportAfterMinutes int
+}
+
+// NewEventReporter returns EventReporter object
+func NewEventReporter(recorder record.EventRecorder, maxCountInGivenTime, reportAfter int) *EventReporter {
+	er := &EventReporter{
+		recorder:                recorder,
+		count:                   maxCountInGivenTime,
+		eventReportAfterMinutes: reportAfter,
+	}
+
+	er.reportedEvents = map[string]*eventObject{}
+
+	return er
+}
+
+// Report records a events if eventReportAfterMinutes has passed or events occurred less than count
+func (rep *EventReporter) Report(instance runtime.Object, eventType, eventReason, msg string) {
+
+	objMeta, err := meta.Accessor(instance)
+
+	if err != nil {
+		return
+	}
+
```

Comment:
Let's go ahead and log the event here, even if it is also logged by the caller. It will give us consistent logging for all the events when events are also added in other places.

Reply:
Done